### ISSUE TITLE
[DNSSEC] resolved gives up on a known good DNS server

### ISSUE THREAD

**Issue by g2p** [systemd/systemd#4315]
### Submission type
- [x] Bug report
- [ ] Request for enhancement (RFE)
### systemd version the issue has been seen with

> 231 (Ubuntu package version 231-9git1)
### Used distribution

> Ubuntu 16.10
### Expected behaviour you didn't see

> DNS names resolve
### Unexpected behaviour you saw

> All names stop resolving after a while, always with the error "incompatible-server"
### Steps to reproduce the problem
1. Enable DNSSEC in /etc/systemd/resolved.conf
2. Use networkctl to confirm that the only resolvers are 8.8.8.8 and 8.8.4.4, which do support DNSSEC
3. Reload systemd-resolved
4. Wait for breakage; takes a few minutes on a busy network
5. Confirm breakage with `systemd-resolve some-domain` or with `journalctl _SYSTEMD_UNIT=systemd-resolved.service`
### Discussion

It seems like the "allow-downgrade" logic is too eager to stop trusting DNS servers.
It detects a broken server when in all likelihood it's just the network being busy and occasionally dropping UDP datagrams.

**Comment by poettering** [comment]
Note that we actually downgrade only if at least 3 DNS queries time out in a row. I am not sure what we can do here. It's hard to distuingish the case of "the DNS server is crap and does not respond to DNSSEC queries at all", and of "your IP connection is extremely lossy".

I figure one could even say: if your connection is _that_ bad then maybe you shouldn't strain it further by the increased traffic DNSSEC generates... Dunno.

Note that if you really want DNSSEC, even over highly lossy links, then simply switch it from DNSSEC=allow-downgrade to DNSSEC=yes, and resolved won't downgrade anymore...

Not sure what else we could do about this. Of course, we can increase the retries, but that means we take even longer before giving up on an actually bad DNS configuration, which will then again upset other folks...

**Comment by g2p** [comment]
I'm using `DNSSEC=yes`. I don't want the downgrade logic, because that logic causes every non-DNSSEC request to fail. Which is surely unintended behaviour.

**Comment by poettering** [comment]
so, what kind of error do you actually get?

(not that we also reduce the non-DNSSEC feature set on 3 lost packets in a row. I.e. we turn off EDNS0 and so on, switch between UDP and TCP and similar things)

**Comment by g2p** [comment]
Here's the error:

```
systemd-resolve test.de
test.de: resolve call failed: DNSSEC validation failed: incompatible-server
```

**Comment by poettering** [comment]
Hmm, so I guess we could say that if the user explicitly set DNSSEC=yes we'll never decide that a server doesn't support DNSSEC because it doesn't respond...

Or maybe we should simply retry for longer in DNSSEC mode if DNSSEC=yes is set before assuming the server is totally crap...

**Comment by g2p** [comment]
I think there's something broken with the detection logic, because it happens almost immediately after I restart systemd-resolved now.
Here's my vote for the first option.

**Comment by poettering** [comment]
@g2p when you use wireshark and make it match on DNS on your server-facing network interface, do you see any answers coming in for the requests resolved is sending?

Do you have any other local DNS service listen on port 53?

**Comment by g2p** [comment]
It took longer to trigger this time (and I was afk), but yes, I get DNS replies before and after the incompatible-server flag is set. There's no other DNS server running.

I have a hard time reading the TCP port 53 part of the wireshark trace, but the moment resolved gives up seems to coincide with FIN,ACK packets on that port.
I didn't get log messages before (because of #1296), but now I get this at the time of the failure:

```
Oct 10 14:50:28 moulinex systemd-resolved[19028]: DNSSEC validation failed for question blah IN SOA: incompatible-server
Oct 10 14:50:28 moulinex systemd-resolved[19028]: DNSSEC validation failed for question bleh IN DS: incompatible-server
Oct 10 14:50:28 moulinex systemd-resolved[19028]: DNSSEC validation failed for question bluh IN SOA: incompatible-server
Oct 10 14:50:28 moulinex systemd-resolved[19028]: DNSSEC validation failed for question blih IN A: incompatible-server
Oct 10 14:50:28 moulinex systemd-resolved[19028]: Using degraded feature set (UDP+EDNS0) for DNS server 8.8.4.4.
```

Before that there was also DO (which I gather is DNSSEC support) in the feature set.

**Comment by poettering** [comment]
if you turn on debug logging for resolved, you should get more info why resolved decides to downgrade. Could you turn that on? (set Environment=SYSTEMD_LOG_LEVEL=debug in the [Service] section of a drop-in /etc/systemd/system/systemd-resolved.service.d/50-debug.conf and issue systemctl daemon-reload and systemctl restart systemd-resolved)

**Comment by g2p** [comment]
I've sent the logs (wireshark capture and journal dump for resolved) by private email.

**Comment by poettering** [comment]
Fix waiting in #5276.
